### PR TITLE
XOR-276 [FEAT] Warn users of live data changes when editing data sources via DSM

### PR DESCRIPTION
**Pull request by NikitaJadhavXoriant**
**Product areas affected**

Studio -> App Data -> Open DS -> Message for live app DS changes

**What does this PR do?**

Implemented functionality to show warning message if app is live and changes made to DS

**JIRA ticket**

[click here](https://weboo.atlassian.net/browse/XOR-276)

**Result**

XOR-276-1

**Checklist**

 None

**Testing instructions**

None

**Deployment instructions**

None

Author concerns

None

### CODE REVIEW

File: js/interface.js [hunk 1]
```diff
@@ -312,8 +316,54 @@ function cacheOriginalEntries(entries, clientIdMap) {
...
+    $('#alert-live-data').html('Some of the data may have been changed by users of the app or other Studio users. Modifying data while live users are accessing the app may overwrite data. We recommend using admin screens within the app to modify data safely. \<a href="#" data-source-reload>Reload\</a> to see the latest version. \<a target="_blank" href="https://help.fliplet.com">Learn more\</a>');
+
+    Fliplet.Studio.emit('track-event', {
+      category: 'dsm_reload_warning',
+      action: 'show'
+    });
+  }, 300000);
+}
+
+/**
```
Comment (amolsuryawanshi47) [inline js/interface.js:344]: Missing JSDoc @returns for function

File: js/interface.js [hunk 2]
```diff
@@ -312,8 +316,54 @@ function cacheOriginalEntries(entries, clientIdMap) {
...
 
+/**
+ * Clear the global timer and hides #alert-live-data by adding a .hidden class
+ */
+function clearLiveDataTimer() {
+  clearTimeout(this.globalTimer);
+  $('#alert-live-data').addClass('hidden');
+}
+
+/**
```
Comment (amolsuryawanshi47) [inline js/interface.js:327]: Missing JSDoc @returns for function

**Review by amolsuryawanshi47 — APPROVED** [review]
changes looks ok to me

**Review by tonytlwu — CHANGES_REQUESTED** [review]
Does not meeting coding standards. ESlint issues not fixed.

**Comment by NikitaJadhavXoriant** [comment]
> Does not meeting coding standards. ESlint issues not fixed.

Coding standards issue resolved.

File: js/interface.js (hunk 1 above)
Comment (NikitaJadhavXoriant) [reply js/interface.js:346]: Added

File: js/interface.js (hunk 2 above)
Comment (NikitaJadhavXoriant) [reply js/interface.js:328]: Added

**Review by amolsuryawanshi47 — APPROVED** [review]

**Review by amolsuryawanshi47 — APPROVED** [review]
Looks ok to me

File: js/interface.js [hunk 3]
```diff
@@ -312,8 +318,57 @@ function cacheOriginalEntries(entries, clientIdMap) {
...
+    });
+  }, 300000);
+}
+
+/**
+ * Checks if current app is published or not
+ * @returns {void}
+ */
+function fetchCurrentDataSourcePublishStatus() {
+  getApps.then(function(apps) {
```
Comment (tonytlwu) [inline js/interface.js:353]: This isn't how we should be determining whether a data source is connected to a live app. I'll update the tech spec with more information.

**Review by tonytlwu — CHANGES_REQUESTED** [review]

File: js/interface.js [hunk 4]
```diff
@@ -73,16 +76,18 @@ var definitionEditor = CodeMirror.fromTextArea($('#definition')[0], {
...
-  appId: widgetData.appId,
-  query: {
-    type: 'integrationToken',
-    order: 'createdAt',
-    direction: 'DESC'
-  }
-}).then(function(tokens) {
-  integrationTokenList = tokens;
-});
+if (widgetData.appId) {
```
[the remark this answers is not in the document]
Comment (NikitaJadhavXoriant) [reply js/interface.js:79]: Hi @tonytlwu , yes you are correct, above code is not related to XOR-276, but after taking latest from master branch of Data Source widget (which involved XOR-203 changes), Kathi found one issue while testing, so to quick fix that issue Amol gave above solution, so I added above code.

File: js/interface.js [hunk 5]
```diff
@@ -312,8 +318,57 @@ function cacheOriginalEntries(entries, clientIdMap) {
...
+  }, 300000);
+}
+
+/**
+ * Checks if current app is published or not
+ * @returns {void}
+ */
+function fetchCurrentDataSourcePublishStatus() {
+  getApps.then(function(apps) {
+    this.currentAppPublishStatus = false;
```
[the remark this answers is not in the document]
Comment (NikitaJadhavXoriant) [reply js/interface.js:354]: Hi @tonytlwu, I was misunderstood the scope. I will remove 'this' and update the PR.

File: js/interface.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (NikitaJadhavXoriant) [reply js/interface.js]: Hi @tonytlwu, we have created new ticket for above issue  #XOR-566, and it's handled separately. So, removing above changes from this PR.

File: js/interface.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (NikitaJadhavXoriant) [reply js/interface.js]: Done changes in PR

File: js/interface.js [hunk 6]
```diff
@@ -312,8 +318,57 @@ function cacheOriginalEntries(entries, clientIdMap) {
...
+
+    var currentApp = _.find(apps, function(app) {
```
[the remark this answers is not in the document]
Comment (NikitaJadhavXoriant) [reply js/interface.js]: Updated code accordingly.

**Review by amolsuryawanshi47 — APPROVED** [review]
Changes looks ok to me

**Review by tonytlwu — CHANGES_REQUESTED** [review]
Does not meet coding standards

File: js/interface.js [hunk 7]
```diff
@@ -273,6 +275,11 @@ function fetchCurrentDataSourceDetails() {
     currentDataSourceRules = dataSource.accessRules;
     currentFinalRules = dataSource.accessRules;
     currentDataSourceDefinition = dataSource.definition || {};
+    if (dataSource.apps && dataSource.apps.length > 0) {
```
Comment (tonytlwu) [inline js/interface.js:278]:
```suggestion

    if (dataSource.apps && dataSource.apps.length > 0) {
```

File: js/interface.js [hunk 8]
```diff
@@ -46,6 +46,8 @@ var currentFinalRules;
 var integrationTokenList;
 var selectedTokenId;
 var selectedTokenName;
+var globalTimer;
+var isPublishedApp = false;
```
Comment (tonytlwu) [inline js/interface.js:50]:
```suggestion
var dataSourceIsLive = false;
```

**Review by tonytlwu — CHANGES_REQUESTED** [review]
The change requests don't need another round of QA

File: js/interface.js (hunk 7 above)
Comment (NikitaJadhavXoriant) [reply js/interface.js:279]: Done

File: js/interface.js (hunk 8 above)
Comment (NikitaJadhavXoriant) [reply js/interface.js]: Done

**Review by amolsuryawanshi47 — APPROVED** [review]
Changes looks ok to me

**Pull request merged**